test/rules/rules.go

```diff
@@ -0,0 +1,15 @@
+//go:build gorules
+
+package rules
+
+import "github.com/quasilyte/go-ruleguard/dsl"
+
+// nolint:unused
+func resourceCleanup(m dsl.Matcher) {
+	m.Match(`$res, $err := $resolver.FileContentsByLocation($*_); if $*_ { $*_ }; $next`).
+		Where(m["res"].Type.Implements(`io.Closer`) &&
+			m["err"].Type.Implements(`error`) &&
+			m["res"].Type.Implements(`io.Closer`) &&
+			!m["next"].Text.Matches(`defer internal.CloseAndLogError`)).
+		Report(`internal.CloseAndLogError should be deferred right after the error returned from $resolver.FileContentsByLocation is checked.`)
```

Comment:
I think this approach is pretty simple to comprehend and the rule defined here makes sense for the specific case that we want to check. I think maybe the report could read something more like:
```
defer internal.CloseAndLogError($res,...) should be called immediately following the error check
```

Reply:
The failing check now looks like this:

```
Error: syft/file/cataloger/executable/cataloger.go:63:3: ruleguard: please call "defer internal.CloseAndLogError(reader, loc.RealPath)" right after checking the error returned from resolver.FileContentsByLocation. (gocritic)
```